tinodesdk/src/main/java/co/tinode/tinodesdk/model/MsgRange.java

```diff
@@ -0,0 +1,176 @@
+        low = id;
+    }
+
+    public MsgRange(int low, int hi) {
+        this.low = low;
+        this.hi = hi;
+    }
+
+    public MsgRange(MsgRange that) {
+        this.low = that.low;
+        this.hi = that.hi;
+    }
+
+    @Override
+    public int compareTo(MsgRange other) {
+        int r = nullableCompare(low, other.low);
+        if (r == 0) {
+            r = nullableCompare(other.hi, hi);
+        }
+        return r;
+    }
+
+    @Override
+    public String toString() {
+        return "{low: " + low + (hi != null ? (", hi: " + hi) : "") + "}";
+    }
+
+    protected boolean tryExtending(int h) {
+        boolean done = false;
+        if (h == low) {
+            done = true;
+        } else if (hi != null && hi != 0) {
+            if (h == hi) {
+                hi ++;
+                done = true;
+            }
+        } else if (h == low + 1) {
+            hi = h + 1;
+            done = true;
+        }
+        return done;
+    }
+
+    // If <b>hi</b> is meaningless or invalid, remove it.
+    protected void normalize() {
+        if (hi != null && hi <= low + 1) {
+            hi = null;
+        }
+    }
+
+    /**
+     * Convert List of IDs to multiple ranges.
+     */
+    public static MsgRange[] listToRanges(final List<Integer> list) {
+        if (list == null || list.size() == 0) {
+            return null;
+        }
+
+        // Make sure the IDs are sorted in ascending order.
+        Collections.sort(list);
+
+        ArrayList<MsgRange> ranges = new ArrayList<>();
+        MsgRange curr = new MsgRange(list.get(0));
+        ranges.add(curr);
+        for (int i = 1; i < list.size(); i++) {
+            if (!curr.tryExtending(list.get(i))) {
+                curr.normalize();
+                // Start a new range
+                curr = new MsgRange(list.get(i));
+                ranges.add(curr);
+            }
+        }
+
+        // No need to sort the ranges. They are already sorted.
+
+        return ranges.toArray(new MsgRange[]{});
+    }
+
+    /**
+     * Collapse multiple possibly overlapping ranges into as few ranges non-overlapping
+     * ranges as possible: [1..6],[2..4],[5..7] -> [1..7].
+     *
+     * The input array of ranges must be sorted.
+     *
+     * @param ranges ranges to collapse
+     * @return non-overlapping ranges.
+     */
+    public static MsgRange[] collapse(MsgRange[] ranges) {
+        if (ranges.length > 1) {
+            int prev = 0;
+            for (int i = 1; i < ranges.length; i++) {
+                if (ranges[prev].low.equals(ranges[i].low)) {
+                    // Same starting point.
+
+                    // Earlier range is guaranteed to be wider or equal to the later range,
+                    // collapse two ranges into one (by doing nothing)
+                    continue;
+                }
+
+                // Check for full or partial overlap
+                int prev_hi = ranges[prev].hi != null ? ranges[prev].hi : ranges[prev].low + 1;
```

Comment:
This seems like a repeated pattern (checking presence of `hi` to get the upper bound of a Range).
How about making it a function in Range?

Reply:
There is a suprizing problem here: I can't name the method `getHi` because of the way jackson xml uses names. Naming it this way will change the serialization and will always include `hi` value into json.  I mean I can but it's wasteful.

Reply:
I added the method and will check how it affects serialization.

Reply:
+1